lib/spack/spack/spec.py

```diff
@@ -1815,38 +1813,32 @@ def node_dict_with_hashes(self, hash=ht.dag_hash):
         node = self.to_node_dict(hash)
         node[ht.dag_hash.name] = self.dag_hash()
 
-        # full_hash and build_hash are lazily computed -- but if we write
-        # a spec out, we want them to be included. This is effectively
-        # the last chance we get to compute them accurately.
+        # dag_hash is lazily computed -- but if we write a spec out, we want it
+        # to be included. This is effectively the last chance we get to compute
+        # it accurately.
         if self.concrete:
-            # build and full hashes can be written out if:
-            # 1. they're precomputed (i.e. we read them from somewhere
-            #    and they were already on the spec
-            # 2. we can still compute them lazily (i.e. we just made them and
+            # dag_hash can be written out if:
+            # 1. it's precomputed (i.e. we read it from somewhere
+            #    and it was already on the spec)
+            # 2. we can still compute it lazily (i.e. we just made the spec and
             #    have the full dependency graph on-hand)
             #
-            # we want to avoid recomputing either hash for specs we read
+            # we want to avoid recomputing the dag_hash for specs we read
```

Comment:
We'll never read a spec from the DB that does not have a DAG hash.  If we do, that's an error.

See my note below in `from_node_dict` for more details.